components/salsa-2022/src/storage.rs

```diff
@@ -155,16 +155,29 @@ where
 
     pub fn ingredient(&self, ingredient_index: IngredientIndex) -> &dyn Ingredient<DB> {
         let route = self.routes.route(ingredient_index);
-        route(&self.shared.jars)
+        route(self.shared.jars.as_ref().unwrap())
+    }
+}
+
+impl<DB> Clone for Shared<DB>
+where
+    DB: HasJars,
+{
+    fn clone(&self) -> Self {
+        Self {
+            jars: self.jars.clone(),
+            cvar: self.cvar.clone(),
+        }
     }
 }
 
-impl<DB> Drop for Shared<DB>
+impl<DB> Drop for Storage<DB>
 where
     DB: HasJars,
 {
     fn drop(&mut self) {
-        self.cvar.notify_all();
+        drop(self.shared.jars.take());
```

Comment:
```suggestion
        // Drop the Arc reference before the cvar is notified,
        // since other threads are sleeping, waiting for it to reach 1.
        drop(self.shared.jars.take());
```